src/net/http/UploadFile.cpp

```diff
@@ -0,0 +1,255 @@
+/*
+Copyright_License {
+
+  XCSoar Glide Computer - http://www.xcsoar.org/
+  Copyright (C) 2000-2021 The XCSoar Project
+  A detailed list of copyright holders can be found in the file "AUTHORS".
+
+  This program is free software; you can redistribute it and/or
+  modify it under the terms of the GNU General Public License
+  as published by the Free Software Foundation; either version 2
+  of the License, or (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
+*/
+
+#include "UploadFile.hpp"
+#include "Request.hxx"
+#include "Handler.hxx"
+#include "Operation/Operation.hpp"
+#include "io/FileOutputStream.hxx"
+#include "thread/Mutex.hxx"
+#include "thread/Cond.hxx"
+#include "util/NumberParser.hpp"
+#include "util/ScopeExit.hxx"
+#include "util/StaticString.hxx"
+#include "util/Exception.hxx"
+#include "time/BrokenDate.hpp"
+#include "Formatter/DateFormatter.hpp"
+#include "system/ConvertPathName.hpp"
+
+#include "LogFile.hpp"
+
+#include <cassert>
+
+class UploadFileHandler final : public CurlResponseHandler {
+  uint8_t *msg_buffer;
+  const size_t max_size;
+  
+  size_t posted = 0;
+  size_t post_count = 0;
+
+  size_t received = 0;
+
+  OperationEnvironment &env;
+
+  Mutex mutex;
+  Cond cond;
+
+  std::exception_ptr error;
+
+  bool done = false;
+  
+ public:
+  UploadFileHandler(void *_msg_buffer, size_t _max_size,
+    OperationEnvironment& _env)
+    :msg_buffer((uint8_t*)_msg_buffer), max_size(_max_size), env(_env) {}
+
+  size_t GetReceived() const {
+    return received;
+  }
+
+  void Cancel() noexcept {
+    const std::lock_guard<Mutex> lock(mutex);
+    done = true;
+    cond.notify_one();
+  }
+
+  void Wait() {
+    std::unique_lock<Mutex> lock(mutex);
+    cond.wait(lock, [this]{ return done; });
+
+    if (error)
+      std::rethrow_exception(error);
+  }
+
+  /* virtual methods from class CurlResponseHandler */
+  void OnHeaders(unsigned /*status*/, std::multimap<std::string,
+      std::string> &&/*headers*/) override {}
+
+  void OnData(ConstBuffer<void> data) override {
+    size_t remaining = max_size - received;
+    if (remaining == 0 || done)
```

Comment:
What should happen if max_size is reached?

Reply:
The message will be cut off at the end of this buffer
Done